File: backend/demand.py
import pandas as pd
import joblib
from flask import Flask, request, jsonify
from flask_cors import CORS

# Load the trained model
model = joblib.load('models/blood_donation_model.pkl')
df = pd.read_csv("blood_demand_data.csv")

# Initialize the Flask app
app = Flask(__name__)
CORS(app)

@app.route('/predict', methods=['POST'])
def predict():
    data = request.get_json()
    try:
        date = data.get("Date")
        population = data.get("Population")
        events = data.get("Events")
        historical_usage = data.get("HistoricalBloodUsage")
        admissions = data.get("HospitalAdmissions")
        donors_available = data.get("BloodDonorsAvailable")
        temperature = data.get("Temperature")

        # Validate date
        if not date:
            return jsonify({"error": "Date is required"}), 400
        date = pd.to_datetime(date)
        day_of_week = date.dayofweek
        month = date.month

        # Validate required fields
        for field in [population, events, historical_usage, admissions, donors_available, temperature]:
            if field is None:
                return jsonify({"error": "All fields are required"}), 400

        input_data = pd.DataFrame([{
            'DayOfWeek': day_of_week,
            'Month': month,
            'Population': population,
            'Events': events,
            'HistoricalBloodUsage': historical_usage,
            'HospitalAdmissions': admissions,
            'BloodDonorsAvailable': donors_available,
            'Temperature': temperature
        }])

        prediction = model.predict(input_data)
        return jsonify({"PredictedBloodDemand": float(prediction[0])})
    except Exception as e:
        print("Prediction error:", e)
        return jsonify({"error": "Prediction failed"}), 500

if __name__ == '__main__':
    app.run(debug=True)

File: backend/routes/demand.js
const express = require('express');
const router = express.Router();

/**
 * This route is a placeholder.
 * You cannot use Python .pkl models directly in Node.js.
 * Instead, use a separate Python backend and call it via /predict.
 */

// GET placeholder
router.get('/', (req, res) => {
  res.json({ message: 'Demand route is not implemented in Node.js. Use /predict for ML.' });
});

// POST placeholder (returns dummy response)
router.post('/', (req, res) => {
  try {
    const { Date: dateStr } = req.body;
    const dateObj = dateStr ? new Date(dateStr) : null;
    res.json({
      message: 'Prediction not available in Node.js. Use /predict route.',
      parsedDate: dateObj ? dateObj.toISOString() : null
    });
  } catch (err) {
    res.status(500).json({ error: 'Prediction failed.' });
  }
});

module.exports = router;
module.exports = router;
      HospitalAdmissions == null ||
      BloodDonorsAvailable == null ||
      Temperature == null
    ) {
      return res.status(400).json({ error: 'All fields are required.' });
    }

    // Convert Date to DayOfWeek and Month
    const dateObj = new Date(Date);
    const dayOfWeek = dateObj.getDay(); // 0=Sunday, 1=Monday, ...
    const month = dateObj.getMonth() + 1; // 1-12

    // Prepare input array
    const inputArr = [
      dayOfWeek,
      month,
      Number(Population),
      Number(Events),
      Number(HistoricalBloodUsage),
      Number(HospitalAdmissions),
      Number(BloodDonorsAvailable),
      Number(Temperature)
    ];

    // ⛔ Node.js cannot load .pkl models
    // Suggest using Python backend (Flask/FastAPI) and call it using fetch/axios
    res.json({
      message: 'Prediction not available in Node.js. Please use /predict route to call Python backend.',
      input: inputArr
    });

  } catch (err) {
    console.error('Demand prediction error:', err);
    res.status(500).json({ error: 'Prediction failed.' });
  }
});

module.exports = router;
